File: app/src/main/java/yoyo/app/android/com/yoyoapp/trip/booking/MyTravellerDialogFragment.java
package yoyo.app.android.com.yoyoapp.trip.booking;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.core.util.Consumer;
import androidx.fragment.app.DialogFragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import yoyo.app.android.com.yoyoapp.DataModels.Traveller;
import yoyo.app.android.com.yoyoapp.R;
import yoyo.app.android.com.yoyoapp.trip.adapter.TravellerCompanionRecyclerviewAddapter;
import java.util.ArrayList;
import java.util.List;


public class MyTravellerDialogFragment extends DialogFragment {

    private RecyclerView recyclerView;
    private TravellerCompanionRecyclerviewAddapter travellerRecyclerviewaddapter;
    private ArrayList<Traveller> travellers;
    private TextView messageTextview;
    private Consumer<Traveller> travellerConsumer;
    private BookingViewModel bookingViewModel;
    private View view;


    public MyTravellerDialogFragment(Consumer<Traveller> travellerConsumer) {
        this.travellerConsumer = travellerConsumer;
    }
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        view = inflater.inflate(R.layout.fragment_my_traveller_dialog, container, false);

        travellers = new ArrayList<>();
        recyclerView = view.findViewById(R.id.rv_mytraveller);
        messageTextview = view.findViewById(R.id.tv_travellerdialog_message);
        bookingViewModel = ViewModelProviders.of(getActivity()).get(BookingViewModel.class);

        getTravellersCompanions();
        setupRecyclerview();

        return view;
    }

    // get traveller companions if the user has them before added
    private void getTravellersCompanions() {

        bookingViewModel.initGetTravellers();
        bookingViewModel.getTravellers().observe(getActivity(), new Observer<List<Traveller>>() {
            @Override
            public void onChanged(List<Traveller> travellerList) {
                if (travellerList != null)
                {
                    if (travellerList.size()!= 0)
                    messageTextview.setVisibility(View.GONE);
                    travellers.addAll(travellerList);
                    travellerRecyclerviewaddapter.notifyDataSetChanged();
                }
            }
        });

    }

    // recycler view for travellers
    private void setupRecyclerview() {
        travellerRecyclerviewaddapter = new TravellerCompanionRecyclerviewAddapter(travellers, getActivity(), new TravellerCompanionRecyclerviewAddapter.OnItemSelected() {
            @Override
            public void onSendResult(Traveller traveller) {
                travellerConsumer.accept(traveller);
                dismiss();
            }
        });
        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(getContext(), RecyclerView.VERTICAL, false);
        recyclerView.setLayoutManager(linearLayoutManager);
        recyclerView.setAdapter(travellerRecyclerviewaddapter);
    }
}

File: app/src/main/java/yoyo/app/android/com/yoyoapp/trip/booking/MyTravellerDialogFragment.kt
package yoyo.app.android.com.yoyoapp.trip.booking

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.viewModels
import androidx.lifecycle.Observer
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import kotlinx.android.synthetic.main.fragment_my_traveller_dialog.view.*
import yoyo.app.android.com.yoyoapp.DataModels.Traveller
import yoyo.app.android.com.yoyoapp.R
import yoyo.app.android.com.yoyoapp.trip.adapter.TravellerCompanionRecyclerviewAddapter
import yoyo.app.android.com.yoyoapp.trip.booking.firstPage.BookingViewModel

import java.util.ArrayList


class MyTravellerDialogFragment(private val f: (Traveller) -> Unit) : DialogFragment() {

    private var travellerRecyclerViewAdapter: TravellerCompanionRecyclerviewAddapter? = null
    private lateinit var travellers: ArrayList<Traveller>
    private val bookingViewModel by viewModels<BookingViewModel>()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val res = inflater.inflate(R.layout.fragment_my_traveller_dialog, container, false)

        travellers = ArrayList()
        getTravellersCompanions()
        setupRecyclerView(res)

        return res
    }

    // get traveller companions if the user has them before added
    private fun getTravellersCompanions() {

        bookingViewModel.initGetTravellers()
        bookingViewModel.travellerCompanionsMutableList.observe(activity!!, Observer { travellerList ->
            if (! travellerList.isNullOrEmpty())
            {
                travellers.addAll(travellerList)
                travellerRecyclerViewAdapter?.notifyDataSetChanged()
            }
                view?.tv_travellerdialog_message?.visibility = View.GONE
        })

    }

    private fun setupRecyclerView(res: View) {
        travellerRecyclerViewAdapter = TravellerCompanionRecyclerviewAddapter(travellers, activity,
            TravellerCompanionRecyclerviewAddapter.OnItemSelected { traveller ->
                f(traveller)
                dismiss()
            })
        val linearLayoutManager = LinearLayoutManager(context, RecyclerView.VERTICAL, false)
        res.rv_mytraveller.layoutManager = linearLayoutManager
        res.rv_mytraveller.adapter = travellerRecyclerViewAdapter
    }
}
